app/services/macros/execution_service.rb

```diff
@@ -0,0 +1,88 @@
+class Macros::ExecutionService
+  def initialize(macro, conversation, user)
+    @macro = macro
+    @account = macro.account
+    @conversation = conversation
+    Current.user = user
+  end
+
+  def perform
+    @macro.actions.each do |action|
+      action = action.with_indifferent_access
+      begin
+        send(action[:action_name], action[:action_params])
+      rescue StandardError => e
+        ChatwootExceptionTracker.new(e, account: @account).capture_exception
+      end
+    end
+  ensure
+    Current.reset
+  end
+
+  private
+
+  def mute_conversation(_params)
+    @conversation.mute!
+  end
+
+  def snooze_conversation(_params)
+    @conversation.snoozed!
+  end
+
+  def resolve_conversation(_params)
+    @conversation.resolved!
+  end
+
+  def change_status(status)
+    @conversation.update!(status: status[0])
+  end
+
+  def send_webhook_event(webhook_url)
+    payload = @conversation.webhook_data.merge(event: "macro_event.#{@macro.name}")
+    WebhookJob.perform_later(webhook_url[0], payload)
+  end
+
+  def send_message(message)
+    return if conversation_a_tweet?
+
+    params = { content: message[0], private: false, content_attributes: { automation_rule_id: @macro.id } }
+    mb = Messages::MessageBuilder.new(nil, @conversation, params)
+    mb.perform
+  end
+
+  def assign_team(team_ids = [])
+    return unless team_belongs_to_account?(team_ids)
+
+    @conversation.update!(team_id: team_ids[0])
+  end
+
+  def assign_best_agent(agent_ids = [])
+    return unless agent_belongs_to_account?(agent_ids)
+
+    @agent = @account.users.find_by(id: agent_ids)
+
+    @conversation.update!(assignee_id: @agent.id) if @agent.present?
+  end
+
+  def add_label(labels)
+    return if labels.empty?
+
+    @conversation.add_labels(labels)
+  end
+
+  def send_email_to_team(_params); end
+
+  def agent_belongs_to_account?(agent_ids)
```

Comment:
some of these methods seem repetitive to what we have in automation service 
https://github.com/chatwoot/chatwoot/blob/c4b2005425f7a02003fb03d80bc45f445a50114b/app/services/automation_rules/action_service.rb#L101

can we extract them to a common base class or a helper and avoid the repetition?

Reply:
Yes, we can have a common action service may be. The issue was just with the activity execution, let me check.